### PR TITLE
Resto druid - added premature rejuvenations module

### LINKED ISSUE #652: [Resto] Add premature rejuvenation applications module
Note: stuff that changes rejuvenation duration: Persistance relic, Amanthul's wisdom, Flourish, legendary wrists, deep rooted. Stuff that applies rejuvenation: 4PT19, Soul of the archdruid, tearstone, PotA

The expected value from this module should be the amount of seconds potentially lost from premature applications, also suggestions if it's too high.

A suggestion: 
* Extend the rejuvenation.js class to track every single rejuvenation of all players. 
* Keep an array of objects of all players with the current duration left on their rejuvenation. 
* Whenever a target has a rejuv event (refreshbuff/applybuff) they get a timestamp, and whenever an extension is made, such as flourish, deep rooted or shoulders extend their duration. 
* Need to take pandemic into consideration too.
* Calculate if the new application of rejuvenation is within the pandemic time frame.
* Need to take into consideration on rejuvenation proccs, maybe have a variable that track proccPrematureApplications and castPrematureApplications (you can't really control proccs).
* Suggestion on the object: 
[Screenshot from Gyazo](https://gyazo.com/3449997fcc9be2a33f14ba92ddec0b3a)

Discuss below if there's something unclear.

### PR DESCRIPTION
**Pull request by buimichael**
Would be nice if someone could double check the logic I have implemented. I've probably made it more complicated than required. 

Fixes #652

### CODE REVIEW

File: src/parser/druid/restoration/modules/features/PrematureRejuvenations.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    healingDone: HealingDone,
+    combatants: Combatants,
+  };
+
+  totalRejuvsCasts = 0;
+  rejuvenations = [];
+  earlyRefreshments = 0;
+  timeLost = 0;
+  pandemicExtensions = 0;
+  pandemicExtensionsTime = 0;
```
Comment (Gebuz) [inline src/parser/druid/restoration/modules/features/PrematureRejuvenations.js]: What is pandemicExtensionsTime needed for when timeLost exists? Is it really relevant to show how much correct pandemic time you had?

File: src/parser/druid/restoration/modules/features/PrematureRejuvenations.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      const pandemicTimestamp = oldRejuv.timestamp + ((REJUV_DURATION * PENDEMIC_THRESHOLD) + MS_BUFFER);
+      if (pandemicTimestamp > event.timestamp) {
+        this.earlyRefreshments++;
+        this.timeLost += pandemicTimestamp - event.timestamp;
+      }
+
+      // Account for pandemic time if hot was applied within the pandemic timeframe.
+      let pandemicTime = 0;
+      if (event.timestamp >= pandemicTimestamp && event.timestamp <= oldRejuv.timestamp + REJUV_DURATION) {
+        pandemicTime = (oldRejuv.timestamp + REJUV_DURATION) - event.timestamp;
```
Comment (Gebuz) [inline src/parser/druid/restoration/modules/features/PrematureRejuvenations.js]: This needs to be capped at 0.3 * rejuv duration

File: src/parser/druid/restoration/modules/features/PrematureRejuvenations.js (hunk 1 above)
Comment (buimichael) [reply src/parser/druid/restoration/modules/features/PrematureRejuvenations.js]: I thought it'd be interesting to show/know that information.

File: src/parser/druid/restoration/modules/features/PrematureRejuvenations.js (hunk 2 above)
Comment (Gebuz) [reply src/parser/druid/restoration/modules/features/PrematureRejuvenations.js]: yeah I misread that if statement. However that means hots that are refreshed early wont be extended by 30% (pandemic time will be 0 if it never enters that if statement).

Comment (buimichael) [reply src/parser/druid/restoration/modules/features/PrematureRejuvenations.js]: Missed that, thanks. Fixed now. @Gebuz

**Pull request merged**